client/Android/FreeRDPCore/jni/android_freerdp.c

```diff
@@ -721,43 +734,70 @@ JNIEXPORT jboolean JNICALL jni_freerdp_disconnect(JNIEnv *env, jclass cls, jint
 	return (jboolean) JNI_TRUE;
 }
 
-JNIEXPORT void JNICALL jni_freerdp_cancel_connection(JNIEnv *env, jclass cls, jint instance)
+JNIEXPORT jboolean JNICALL jni_freerdp_cancel_connection(JNIEnv *env, jclass cls, jint instance)
 {
-	jni_freerdp_disconnect(env, cls, instance);
+	return jni_freerdp_disconnect(env, cls, instance);
 }
 
-JNIEXPORT void JNICALL jni_freerdp_set_data_directory(JNIEnv *env, jclass cls, jint instance, jstring jdirectory)
+JNIEXPORT jboolean JNICALL jni_freerdp_set_data_directory(JNIEnv *env, jclass cls, jint instance, jstring jdirectory)
 {
 	freerdp* inst = (freerdp*)instance;
 	rdpSettings * settings = inst->settings;
 
-	const jbyte *directory = (*env)->GetStringUTFChars(env, jdirectory, NULL);
+	const jbyte* directory = (*env)->GetStringUTFChars(env, jdirectory, NULL);
+	if (!directory)
+		return JNI_FALSE;
 
 	free(settings->HomePath);
 	free(settings->ConfigPath);
```

Comment:
@nfedera settings get freed properly on shutdown so I didn't handle any setting values if not required.